### DISCUSSION TITLE
Suggested improvement to the "Group by annotation color, customized label" cheat sheet

### DISCUSSION THREAD

**Discussion by FeralFlora** [aidenlx/obsidian-zotero#178] [Ideas]
I suggest changing the for-loop that outputs the headings and associated annotations from this:

```
for (const color in byColor) { -%>

## <%= label[color] ?? color %>
  <%_ for (const annot of byColor[color]) { %>

<%~ include("annotation", annot) %>
  <%_ } %>
<% } %>
```

to this:

```
for (const color in label) { -%>

<% if (color in byColor) { %>
### <%= label[color] ?? color %>
  <%_ for (const annot of byColor[color]) { %>

<%~ include("annotation", annot) %>
    <%_ } %>
  <%_ } %>
<% } %>
```

I loop over the colors in `label` rather than `byColor`, and add a conditional so that colors in `byColor` are output if they are present in label. This way, the color group headings will be output in the same order as set in label. This ensures that literature notes are consistently formatted. Without this, the groups will be in the sequence of colors in the source document, so it will not be consistent.

There might need to be some else statement to catch colors that are not in `label`.

**Comment by FeralFlora** [comment]
I have tested this, and it works.

**Comment by aidenlx** [comment]
Here is the updated template: 
```ejs
<% const byColor = Object.groupBy(it, (annot) => annot.colorName);
const label = {
  "red": "Important",
  "orange": "Question",
  "yellow": "Summary",
  "gray": "Comment",
  "green": "Answer",
  "cyan": "Task",
  "blue": "Fact",
  "navy": "Definition",
  "purple": "Quote",
  "brown": "Source",
  "magenta": "To Do",
};
// Merge colors with customized label with unexpected colors, if any
// Keep the order of the colors from the original color-label map
const colorSet = new Set([...Object.keys(label), ...Object.keys(byColor)]);
for (const color of colorSet) { 
if (!(color in byColor)) continue -%>

### <%= label[color] ?? color %>
  <%_ for (const annot of byColor[color]) { %>

<%~ include("annotation", annot) %>
  <%_ } %>
<% } %>
```

**Comment by aidenlx** [comment]
Thanks for this — the ordering fix went into the v1 cheat sheet at the time.

In v2's Liquid templates, deterministic heading order is just how you'd write it, rather than a workaround: loop your own label order and pull each group with `where`.

```liquid
{% assign order = "red,orange,yellow,green,blue,purple,magenta,gray" | split: "," %}
{% for color in order %}
{% assign group = zt.annotations | where: "colorName", color %}
{% if group.size > 0 %}
### {{ color }}
{% for annotation in group %}{% render "annotation" with annotation as zt %}{% endfor %}
{% endif %}
{% endfor %}
```

v2 beta is available now: https://zotlit.aidenlx.site/docs/install-zotlit
Migration guide: https://zotlit.aidenlx.site/docs/how-to/migrate-from-v1
